src/passes/TypeMerging.cpp

```diff
@@ -88,188 +91,331 @@ struct CastFinder : public PostWalker<CastFinder> {
+        }
+        return getTempTupleType(vec);
+      };
+
+      auto oldSig = oldSignatureType.getSignature();
+      sig.params = getUpdatedTypeList(oldSig.params);
+      sig.results = getUpdatedTypeList(oldSig.results);
+    }
+  } rewriter(*module, merges);
+}
 
-    class TypeInternalsUpdater : public GlobalTypeRewriter {
-      const TypeUpdates& merges;
+bool TypeMerging::mayBeMergeable(HeapType sub, HeapType super) {
+  // If the type is distinguishable from its supertype, we cannot merge it.
+  if (castTypes.count(sub)) {
+    return false;
+  }
+  // Check whether `sub` and `super` have the same top-level structure,
+  // including the position and identity of any children that are not included
+  // as transitions in the DFA, i.e. any children that are not nontrivial
+  // references.
+  if (sub.isStruct() && super.isStruct()) {
+    return mayBeMergeable(sub.getStruct(), super.getStruct());
+  }
+  if (sub.isArray() && super.isArray()) {
+    return mayBeMergeable(sub.getArray(), super.getArray());
+  }
+  if (sub.isSignature() && super.isSignature()) {
+    return mayBeMergeable(sub.getSignature(), super.getSignature());
+  }
+  return false;
+}
 
-      std::unordered_map<HeapType, Signature> newSignatures;
+bool TypeMerging::mayBeMergeable(const Struct& a, const Struct& b) {
+  if (a.fields.size() != b.fields.size()) {
+    return false;
+  }
+  for (size_t i = 0; i < a.fields.size(); ++i) {
+    if (!mayBeMergeable(a.fields[i], b.fields[i])) {
+      return false;
+    }
+  }
+  return true;
+}
 
-    public:
-      TypeInternalsUpdater(Module& wasm, const TypeUpdates& merges)
-        : GlobalTypeRewriter(wasm), merges(merges) {
+bool TypeMerging::mayBeMergeable(Array a, Array b) {
+  return mayBeMergeable(a.element, b.element);
+}
 
-        // Map the types of expressions (curr->type, etc.) to their merged
-        // types.
-        mapTypes(merges);
+bool TypeMerging::mayBeMergeable(Signature a, Signature b) {
+  return mayBeMergeable(a.params, b.params) &&
+         mayBeMergeable(a.results, b.results);
+}
 
-        // Update the internals of types (struct fields, signatures, etc.) to
-        // refer to the merged types.
-        update();
-      }
+bool TypeMerging::mayBeMergeable(Field a, Field b) {
+  return a.packedType == b.packedType && mayBeMergeable(a.type, b.type);
+}
 
-      Type getNewType(Type type) {
-        if (!type.isRef()) {
-          return type;
-        }
-        auto heapType = type.getHeapType();
-        auto iter = merges.find(heapType);
-        if (iter != merges.end()) {
-          return getTempType(Type(iter->second, type.getNullability()));
-        }
-        return getTempType(type);
-      }
+bool TypeMerging::mayBeMergeable(Type a, Type b) {
+  if (a.isTuple() && b.isTuple()) {
+    return mayBeMergeable(a.getTuple(), b.getTuple());
+  }
+  return a == b || (a.isRef() && b.isRef() && !a.getHeapType().isBasic() &&
+                    !b.getHeapType().isBasic());
+}
```

Comment:
In most cases similar to this, `foo(Type ..)` ends up calling `foo(HeapType ..)` on references, that is, it looks through the type (and possibly handles nullability, if necessary). Why does this not do that?

Reply:
Good point, I think that ignoring the nullability here is a bug. Ignoring the non-basic heap type is intentional, though. We don't know which non-basic heap types are going to end up getting merged, so when creating partitions we ignore differences between non-basic heap type children and the let the DFA state partitioning algorithm take care of differentiating between them.

Reply:
I see, makes sense. Please add a comment though, at least for me that is not obvious.

Reply:
There is a comment in the latest code 👍